docs/designs/link-output.md

```diff
@@ -0,0 +1,97 @@
+
+- Absolute link in source content will NOT processed by docfx
+
+| Links in source content | Examples              | Output in `links.json` |
+| ----------------------- | --------------------- | ---------------------- |
+| Relative Link           | ./a.md (valid)        | /basepath/article/a    |
+|                         | a/b.md (invalid)      | a/b.md                 |
+| Absolute Link           | /basepath/articles/b  | /basepath/articles/b   |
+|                         | www.github.com/x      | www.github.com/x       |
+|                         | https://docs.com/y    | https://docs/com/y     |
+
+> NOTE: the absolute links converted from relative links by docfx don't contains `locale` and `versioning` information
+
+##  Bookmarks
+
+`bookmarks` includes all bookmarks ids generated/defined from source content, includes its inclusions(token/code-snippets)
+
+| Source content examples        | Html content                             | Output in `links.json` |
+| ------------------------------ | -----------------------------------------| ---------------------- |
+|  `# Title H1`                  | `<h1 id="title-h1">Title H1</h1>`        | title-h1               |
+|  `## Title H2`                 | `<h2 id="title-h2">Title H2</h2>`        | title-h2               |
+|  `## Title H2`                 | `<h2 id="title-h2">Title H2</h2>`        | title-h2               |
+| `<a id="##special-chars"></a>` | `<a id="##special-chars"></a>`           | ##special-chars        |
+| `<a id="with:colon"></a>`      | `<a id="with:colon"></a>`                | with:colon             |
+| `[!include[](d.md)]`           | `<h1 id="title-h1">Title H1</h1>`        | title-h1               |
+
+> The content in d.md is `# Title H1` in above case
+
+## Output
+
+`links.json` file will be generated by docfx to contains all the links and bookmarks information about all the page files in build.
+
+The output is **an array** of below data model
+
+  |            |                      |
+  |----------- |--------------------- |
+  | `path`     | dotnet/articles/a.md |
+  | `site-url` | /dotnet/articles/a   |
+  | `monikers` | ['dotnet-1.1']       |
+  | `links`    | *link model*         |
+  | `bookmarks`| *bookmark model*     |
+
+*link model*
+
+  |            |                      |
+  |----------- |--------------------- |
+  | `link`     | dotnet/articles/b |
+
+*bookmark model*
+
+  |            |                      |
+  |----------- |--------------------- |
+  | `id`       | title-h1             |
+
+Example:
+
+```json
+[
+    {
+        "path": "dotnet/articles/a.md",
+        "site-url": "/dotnet/articles/a",
+        "links": 
+        [
+            {
+                "link": "dotnet/articles/b"
+            },
+            {
+                "link": "./c.md"
+            },
+            {
+                "link": "https://docs.com/y"
+            }
+        ],
+        "bookmarks": 
```

Comment:
I'm not sure if we can simply `bookmarks` into an array, link is an object because we might very much likely add additional properties like link type, validate or not. I cannot currently think of a useful property that is tied to a bookmark.
```suggestion
        "bookmarks": [ "title-h1", "title-2" ]
```

Reply:
`links` can also be an array?

---
In reply to: [321062224](https://github.com/dotnet/docfx/pull/5062#discussion_r321062224) [](ancestors = 321062224)

Reply:
@herohua , for extensionality. we want to make it as object in case we want to add more properties like `type`